### PR TITLE
ARROW-9956: [C++] [Gandiva] Implementation of binary_string function in gandiva

### PR DESCRIPTION
**Pull request by naman1996**
Function takes in a normal string or a hexadecimal encoded string( Eg: \xca\xfe\xba\xbe) and converts it to VARBINARY (byte array).

### CODE REVIEW

**Comment by wesm** [comment]
We have hex encoding / decoding functions in arrow/util/string.h, is it possible for you to use them?

**Comment by naman1996** [comment]
> We have hex encoding / decoding functions in arrow/util/string.h, is it possible for you to use them?

I think I can use the arrow::ParseHexValue from string util. This would require a lot to type conversions though as that function takes in a string and returns the hex value an unsigned int.

**Comment by naman1996** [comment]
> We have hex encoding / decoding functions in arrow/util/string.h, is it possible for you to use them?

@wesm I have refactored the code to use arrow::ParseHexValue.

**Comment by naman1996** [comment]
Sorry was closed by mistake 😓

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -835,6 +835,48 @@ const char* replace_utf8_utf8_utf8(gdv_int64 context, const char* text,
...
 }
 
+FORCE_INLINE
+const char* binary_string(gdv_int64 context, const char* text, gdv_int32 text_len,
+                          gdv_int32* out_len) {
+  gdv_binary ret =
+      reinterpret_cast<gdv_binary>(gdv_fn_context_arena_malloc(context, text_len));
+
+  if (ret == nullptr) {
+    gdv_fn_context_set_error_msg(context, "Could not allocate memory for output string");
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/string_ops.cc:845]: function that can return errors needs to set that in the function definition : NativeFunction::kNeedsContext | NativeFunction::kCanReturnErrors

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 2]
```diff
@@ -835,6 +835,48 @@ const char* replace_utf8_utf8_utf8(gdv_int64 context, const char* text,
...
+    *out_len = 0;
+    return "";
+  }
+
+  // converting hex encoded string to normal string
+  int j = 0;
+  for (int i = 0; i < text_len; i++, j++) {
+    if (text[i] == '\\' && i + 3 < text_len &&
+        (text[i + 1] == 'x' || text[i + 1] == 'X')) {
+      std::string hex_string;
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/string_ops.cc:860]: same as before - please avoid use of std::string

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 1 above)
Comment (naman1996) [reply cpp/src/gandiva/precompiled/string_ops.cc:845]: added

Comment (praveenbingo) [reply cpp/src/gandiva/precompiled/string_ops.cc:845]: i still see the same above?

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -835,6 +835,51 @@ const char* replace_utf8_utf8_utf8(gdv_int64 context, const char* text,
...
+  int j = 0;
+  for (int i = 0; i < text_len; i++, j++) {
+    if (text[i] == '\\' && i + 3 < text_len &&
+        (text[i + 1] == 'x' || text[i + 1] == 'X')) {
+      char hex_string[3];
+      hex_string[0] = toupper(text[i + 2]);
+      hex_string[1] = toupper(text[i + 3]);
+      hex_string[2] = '\0';
+      uint8_t out;
+      arrow::Status st;
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/string_ops.cc:865]: have we used this in other places..are we sure it does not link in any streams which might cause issues..

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 4]
```diff
@@ -835,6 +835,51 @@ const char* replace_utf8_utf8_utf8(gdv_int64 context, const char* text,
...
+      st = arrow::ParseHexValue(hex_string, &out);
+      st = arrow::ParseHexValue(hex_string, &out);
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/string_ops.cc:867]: why are we doing this twice..

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 1 above)
Comment (naman1996) [reply cpp/src/gandiva/precompiled/string_ops.cc:845]: Resolving based on offline discussion.

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 3 above)
Comment (naman1996) [reply cpp/src/gandiva/precompiled/string_ops.cc:865]: Not used in other places.
 I have run Dremio unit tests which were using this function on macOS. For linux a Gerrit build is currently running which should run unit tests which must us this function. So should be good IMO.

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 4 above)
Comment (naman1996) [reply cpp/src/gandiva/precompiled/string_ops.cc]: just a typo during refactor. Changed it

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 3 above)
Comment (naman1996) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Checked the function. They are including string. So I have removed it's usages.

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 5]
```diff
@@ -84,6 +83,15 @@ VAR_LEN_OP_TYPES(BINARY_RELATIONAL, greater_than_or_equal_to, >=)
   INNER(NAME, utf8)                \
   INNER(NAME, binary)
 
+int to_binary_from_hex(char ch) {
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/string_ops.cc:86]: inline..is this only to handle ascii characters?

Comment (naman1996) [reply cpp/src/gandiva/precompiled/string_ops.cc:86]: will handle only valid hexadecimal characters
a-f,  A-F,  0 - 9

There is a check too before invoking this method _isxdigit(hd1) && isxdigit(hd2)_

**Review by praveenbingo — APPROVED** [review]

**Pull request closed without merging**